### PR TITLE
[WIP]: Prepare fastboot addon for Fasboot 1.0 release

### LINKED ISSUE #292: EmberApp requires defaults as first argument, otherwise: JavaScript heap out of memory
Hello! I’m excited about this. I did some preliminary investigation on how much work it would be to get `travis-web` FastBoot-ing and I quickly ran into an error I didn’t know how to move beyond: `JavaScript heap out of memory`. I’ve put the error at the bottom due to weird GFM problems.

Thanks to @danmcclain and @stefanpenner in the Ember Community Slack, I was able to get past this with [this commit](https://github.com/travis-ci/travis-web/commit/f1302c8e36c4ba2b2f48ba4bc72b57985924b92f) to add `defaults` as the first argument for `EmberApp`.

I see that this is now [the default](https://github.com/ember-cli/ember-cli/blob/7a60e70356e9af566876905bb271bf568e43e687/blueprints/app/files/ember-cli-build.js) in the `ember-cli-build.js` blueprint, so this won’t be a problem for most people. But it could be something others run into when working on this with older Ember CLI applications. @danmcclain mentioned that @arjansingh had this problem too.

Maybe it’s not worth documenting this at all, as I’m not sure how rare this kind of thing is, but at least this issue may help others who are searching for how to handle this.

<details>
<summary>Stack trace from <code>ember fastboot --serve-assets</code></summary>
<pre>
<--- Last few GCs --->

  128427 ms: Mark-sweep 1318.9 (1404.0) -> 1318.9 (1405.0) MB, 405.0 / 0.0 ms [allocation failure] [GC in old space requested].
  128891 ms: Mark-sweep 1318.9 (1405.0) -> 1318.8 (1405.0) MB, 463.7 / 0.0 ms [allocation failure] [GC in old space requested].
  129587 ms: Mark-sweep 1318.8 (1405.0) -> 1326.1 (1404.0) MB, 695.6 / 0.0 ms [last resort gc].
  130250 ms: Mark-sweep 1326.1 (1404.0) -> 1333.7 (1404.0) MB, 662.9 / 0.0 ms [last resort gc].

<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2343144cfb51 <JS Object>
    2: _compile [module.js:~510] [pc=0x3c9fcfb5e446] (this=0xbc0d1a7da49 <a Module with map 0x1f3b6811c3b1>,content=0x40381104201 <Very long string[648478]>,filename=0x193e69ebfc1 <String[86]: /Users/b/Documents/Travis/travis-web/bower_components/ember/ember-template-compiler.js>)
    3: .js [module.js:~577] [pc=0x3c9fcfb85cb2] (this=0x326b70306391 <an Object with map 0x1f3b6811b121>,module=0xb...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 4: v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 5: v8::internal::String::SlowFlatten(v8::internal::Handle<v8::internal::ConsString>, v8::internal::PretenureFlag) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 6: v8::internal::Parser::ParseProgram(v8::internal::Isolate*, v8::internal::ParseInfo*) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 7: v8::internal::Parser::Parse(v8::internal::ParseInfo*) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 8: v8::internal::Parser::ParseStatic(v8::internal::ParseInfo*) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
 9: v8::internal::(anonymous namespace)::CompileToplevel(v8::internal::CompilationInfo*) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
10: v8::internal::Compiler::GetSharedFunctionInfoForScript(v8::internal::Handle<v8::internal::String>, v8::internal::Handle<v8::internal::Object>, int, int, v8::ScriptOriginOptions, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Context>, v8::Extension*, v8::internal::ScriptData**, v8::ScriptCompiler::CompileOptions, v8::internal::NativesFlag, bool) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
11: v8::ScriptCompiler::CompileUnboundInternal(v8::Isolate*, v8::ScriptCompiler::Source*, v8::ScriptCompiler::CompileOptions, bool) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
12: node::ContextifyScript::New(v8::FunctionCallbackInfo<v8::Value> const&) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
13: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
14: v8::internal::Builtin_Impl_HandleApiCallConstruct(v8::internal::(anonymous namespace)::BuiltinArguments<(v8::internal::BuiltinExtraArguments)1>, v8::internal::Isolate*) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
15: v8::internal::Builtin_HandleApiCallConstruct(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/opt/nvm/versions/node/v6.9.1/bin/node]
16: 0x3c9fc8c092a7
[1]    48705 abort      ember fastboot --serve-assets
</pre>
</details>

### PR DESCRIPTION
**Pull request by kratiahuja**
In order to unblock Fastboot 1.0 release, we want to make the fastboot build performant. We will achieve this by creating additive assets (see this [issue](https://github.com/ember-cli/ember-cli/issues/6350) for details). This PR primarily addresses some of the work already. 

This PR primarily addresses the following issues:
- [x] Double fastboot builds [issue](https://github.com/ember-fastboot/ember-cli-fastboot/issues/264). Corresponding ember-cli changes are in this [PR] (https://github.com/ember-cli/ember-cli/pull/6395) and [RFC](https://github.com/ember-cli/rfcs/pull/77)
- [x] Removes the monkey patching introduced with fastboot build. Related [issue](https://github.com/ember-fastboot/ember-cli-fastboot/issues/276)
- [x] Updates generated package.json to take in an array of app files (containing app.js and app-fastboot.js). Addresses this [issue](https://github.com/ember-fastboot/ember-cli-fastboot/issues/270)
- [x] Addresses the compatibility story around how to keep this a backward compatible change.
- [x] Corresponding fastboot lib changes are [here](https://github.com/ember-fastboot/fastboot/pull/97)
- [x] Fixes #292

TODO:
- [ ] Cleanup and tidy this work
- [ ] Add more comments to explain what this PR is doing
- [ ] Make fastboot app work using `ember-serve`. Corresponding [issue](https://github.com/ember-fastboot/ember-cli-fastboot/issues/274).
- [ ] Fix tests once ember-cli changes land and this [issue](https://github.com/ember-fastboot/ember-cli-fastboot/issues/274) is fixed.
- [ ] Update documentation
- [ ] Kill `ember fastboot` command
    
cc: @tomdale @stefanpenner @rwjblue @kellyselden @ryanone @danmcclain @arjansingh @ryanone @masonwan @kristoferbaxter

### CODE REVIEW

**Comment by kratiahuja** [comment]
Here is a [gist](https://gist.github.com/kratiahuja/e8f7c80a49deebc86a6b7130ef6d4764) of how this can run with fastboot currently.

File: index.js [hunk 1]
```diff
@@ -2,63 +2,112 @@
...
+var existsSync = require('exists-sync');
 
 var EventEmitter = require('events').EventEmitter;
 var mergeTrees = require('broccoli-merge-trees');
 var VersionChecker = require('ember-cli-version-checker');
 
-var patchEmberApp     = require('./lib/ext/patch-ember-app');
 var fastbootAppModule = require('./lib/utilities/fastboot-app-module');
 
 var filterInitializers = require('fastboot-filter-initializers');
```
Comment (ryanone) [inline index.js]: Is this dependency needed still?

Comment (kratiahuja) [inline index.js]: Nope, this was the dirty ember-app private patching that we was causing issues and perf problems.

Comment (ryanone) [inline index.js]: I was actually referring to fastboot-filter-initializers :)

Comment (kratiahuja) [inline index.js]: Aah no i need to remove that.

**Comment by kratiahuja** [comment]
Going to close this since there is no clear decision yet on how we want to approach the build problem. We can reopen this whenever a decision is made.

**Pull request closed without merging**